lib/utils/aws/organizations/matchingaccounts.go

```diff
@@ -0,0 +1,288 @@
+	// If contains "*", all OUs are excluded.
+	// Optional.
+	ExcludeOUs []string
+}
+
+func (m *MatchingAccountsFilter) checkAndSetDefaults() error {
+	if m.OrganizationID == "" {
+		return trace.BadParameter("OrganizationID is required")
+	}
+
+	if len(m.IncludeOUs) == 0 {
+		m.IncludeOUs = []string{"*"}
+	}
+
+	if len(m.IncludeOUs) > 1 && slices.Contains(m.IncludeOUs, "*") {
+		return trace.BadParameter("IncludeOUs cannot contain '*' along with other OU IDs")
+	}
+
+	if len(m.ExcludeOUs) > 1 && slices.Contains(m.ExcludeOUs, "*") {
+		return trace.BadParameter("ExcludeOUs cannot contain '*' along with other OU IDs")
+	}
+
+	return nil
+}
+
+func (m *MatchingAccountsFilter) isIncludeNothing() bool {
+	return len(m.ExcludeOUs) == 1 && m.ExcludeOUs[0] == "*"
+}
+
+func (m *MatchingAccountsFilter) isIncludeAll() bool {
+	return len(m.ExcludeOUs) == 0 && (len(m.IncludeOUs) == 0 || (len(m.IncludeOUs) == 1 && m.IncludeOUs[0] == "*"))
+}
+
+// MatchingAccounts returns the list of account IDs that are part of the organization and match the filter.
+// Every OU in ExcludeOUs is excluded from the results, including its children OUs.
+func MatchingAccounts(ctx context.Context, orgsClient OrganizationsClient, filter MatchingAccountsFilter) ([]string, error) {
+	if err := filter.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if filter.isIncludeNothing() {
+		return []string{}, nil
+	}
+
+	if filter.isIncludeAll() {
+		accountIDs, err := allAccounts(ctx, orgsClient, filter.OrganizationID)
+		return accountIDs, trace.Wrap(err)
+	}
+
+	// General case: build the organization tree, this tree already removes all the excluded OUs and their children OUs.
+	orgTree, err := buildOrgTree(ctx, orgsClient, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Then: iterate over the tree and collect the included accounts.
+	includeEverything := len(filter.IncludeOUs) == 0 || filter.IncludeOUs[0] == "*"
+	return collectIncludedAccounts(orgTree, filter.IncludeOUs, includeEverything), nil
+}
+
+func collectIncludedAccounts(orgItem *awsOrgItem, included []string, isParentIncluded bool) []string {
+	includeCurrentOU := isParentIncluded || slices.Contains(included, orgItem.id)
+
+	var accountIDs []string
+	if includeCurrentOU {
+		accountIDs = append(accountIDs, orgItem.accounts...)
+	}
+
+	for _, orgUnit := range orgItem.organizationalUnits {
+		childAccountIDs := collectIncludedAccounts(orgUnit, included, includeCurrentOU)
+		accountIDs = append(accountIDs, childAccountIDs...)
+	}
+
+	return accountIDs
+}
+
+func allAccounts(ctx context.Context, orgsClient OrganizationsClient, organizationID string) ([]string, error) {
+	var accountIDs []string
+	organizationIDValidated := false
+
+	paginator := organizations.NewListAccountsPaginator(orgsClient, &organizations.ListAccountsInput{})
+
+	if paginator.HasMorePages() {
+		page, err := paginator.NextPage(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
```

Comment:
am i reading this right, or does this only process one page? im trying to find where we loop through the rest of the pages if `HasMorePages` continues to be true

Reply:
Nice catch, thank you.